torchtune/models/llama3/_tokenizer.py

```diff
@@ -166,11 +166,34 @@ def decode(
         Returns:
             str: The decoded string.
         """
-        return self.tt_model.decode(
-            token_ids,
-            truncate_at_eos=truncate_at_eos,
-            skip_special_tokens=skip_special_tokens,
-        )
+        if skip_special_tokens:
+            not_header = [True] * len(token_ids)
+            mask = True
+            idx = 0
+            while idx < len(token_ids):
+                if token_ids[idx] == self.start_header_id:
+                    # Start masking out at beginning of header
+                    mask = False
+                elif token_ids[idx] == self.end_header_id:
+                    # Mask out end header id and "\n\n" after it, then reset mask
+                    not_header[idx] = False
+                    mask = True
+                    idx += 1
+                    continue
+                not_header[idx] = mask
+                idx += 1
+
+            return self.tt_model.decode(
+                [token_ids[i] for i, m in enumerate(not_header) if m],
```

Comment:
I suggested defining token_ids = [token_ids[i] for i, m in enumerate(not_header) if m]

and having a single return outside, so we can remove the if/else and duplicated code